File: padlock/src/main/java/com/pyamsoft/padlock/list/LockListItem.java
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.list;

import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v4.app.FragmentActivity;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.CompoundButton;
import com.mikepenz.fastadapter.items.GenericAbstractItem;
import com.pyamsoft.padlock.Injector;
import com.pyamsoft.padlock.R;
import com.pyamsoft.padlock.databinding.AdapterItemLocklistBinding;
import com.pyamsoft.padlock.uicommon.AppIconLoader;
import com.pyamsoft.padlock.model.AppEntry;
import com.pyamsoft.pydroid.loader.ImageLoader;
import com.pyamsoft.pydroid.loader.LoaderMap;
import com.pyamsoft.pydroid.loader.loaded.Loaded;
import com.pyamsoft.pydroid.ui.util.DialogUtil;
import java.util.List;
import javax.inject.Inject;
import timber.log.Timber;

public class LockListItem
    extends GenericAbstractItem<AppEntry, LockListItem, LockListItem.ViewHolder>
    implements FilterableItem<LockListItem, LockListItem.ViewHolder> {

  @NonNull private final LoaderMap loaderMap = new LoaderMap();
  @SuppressWarnings("WeakerAccess") @Inject LockListItemPresenter presenter;

  LockListItem(@NonNull AppEntry entry) {
    super(entry);
    Injector.get().provideComponent().inject(this);
  }

  @Override public int getType() {
    return R.id.adapter_lock_item;
  }

  @Override public int getLayoutRes() {
    return R.layout.adapter_item_locklist;
  }

  @Override public ViewHolder getViewHolder(View view) {
    return new ViewHolder(view);
  }

  @Override public boolean filterAgainst(@NonNull String query) {
    final String name = getModel().name().toLowerCase().trim();
    Timber.d("Filter predicate: '%s' against %s", query, name);
    return !name.startsWith(query);
  }

  @Override public void bindView(ViewHolder holder, List<Object> payloads) {
    super.bindView(holder, payloads);
    holder.binding.lockListTitle.setText(getModel().name());
    holder.binding.lockListToggle.setOnCheckedChangeListener(null);
    holder.binding.lockListToggle.setChecked(getModel().locked());

    Loaded appIcon = ImageLoader.fromLoader(AppIconLoader.forPackageName(getModel().packageName()))
        .into(holder.binding.lockListIcon);
    loaderMap.put("locked", appIcon);

    holder.binding.lockListToggle.setOnCheckedChangeListener(
        new CompoundButton.OnCheckedChangeListener() {
          @Override public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
            buttonView.setOnCheckedChangeListener(null);
            buttonView.setChecked(!isChecked);

            final CompoundButton.OnCheckedChangeListener listener = this;
            presenter.modifyDatabaseEntry(isChecked, getModel().packageName(), null,
                getModel().system(), new LockListItemPresenter.DatabaseCallback() {

                  @Override public void onDatabaseEntryError() {
                    DialogUtil.onlyLoadOnceDialogFragment(
                        (FragmentActivity) holder.itemView.getContext(), new ErrorDialog(),
                        "error");
                  }

                  private void updateModel(boolean locked) {
                    withModel(getModel().toBuilder().locked(locked).build());
                  }

                  @Override public void onDatabaseEntryCreated() {
                    updateModel(true);
                    buttonView.setChecked(true);
                    buttonView.setOnCheckedChangeListener(listener);
                  }

                  @Override public void onDatabaseEntryDeleted() {
                    updateModel(false);
                    buttonView.setChecked(false);
                    buttonView.setOnCheckedChangeListener(listener);
                  }
                });
          }
        });
  }

  @Override public void unbindView(ViewHolder holder) {
    super.unbindView(holder);
    holder.binding.lockListTitle.setText(null);
    holder.binding.lockListIcon.setImageDrawable(null);
    holder.binding.lockListToggle.setOnCheckedChangeListener(null);

    presenter.stop();
    presenter.destroy();

    loaderMap.clear();
  }

  static final class ViewHolder extends RecyclerView.ViewHolder {

    @NonNull final AdapterItemLocklistBinding binding;

    ViewHolder(View itemView) {
      super(itemView);
      binding = DataBindingUtil.bind(itemView);
    }
  }
}

File: padlock/src/main/java/com/pyamsoft/padlock/list/LockListItem.kt
/*
 * Copyright 2019 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.pyamsoft.padlock.list

import android.view.View
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LifecycleRegistry
import androidx.recyclerview.widget.RecyclerView
import com.mikepenz.fastadapter.items.ModelAbstractItem
import com.pyamsoft.padlock.Injector
import com.pyamsoft.padlock.PadLockComponent
import com.pyamsoft.padlock.R
import com.pyamsoft.padlock.model.list.AppEntry
import com.pyamsoft.pydroid.core.bus.EventBus
import com.pyamsoft.pydroid.core.bus.Publisher
import com.pyamsoft.pydroid.util.fakeBind
import com.pyamsoft.pydroid.util.fakeUnbind
import timber.log.Timber
import javax.inject.Inject

class LockListItem internal constructor(
  internal var activity: FragmentActivity,
  entry: AppEntry
) : ModelAbstractItem<AppEntry, LockListItem, LockListItem.ViewHolder>(entry),
    FilterableItem<LockListItem, LockListItem.ViewHolder> {

  override fun getType(): Int = R.id.adapter_lock_item

  override fun getLayoutRes(): Int = R.layout.adapter_item_locklist

  override fun getViewHolder(view: View): ViewHolder = ViewHolder(view)

  override fun filterAgainst(query: String): Boolean {
    val name = model.name.toLowerCase()
        .trim { it <= ' ' }
    Timber.d("Filter predicate: '%s' against %s", query, name)
    return name.contains(query)
  }

  override fun bindView(
    holder: ViewHolder,
    payloads: List<Any>
  ) {
    super.bindView(holder, payloads)
    holder.bind(model)
  }

  override fun unbindView(holder: ViewHolder) {
    super.unbindView(holder)
    holder.unbind()
  }

  class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView),
      LifecycleOwner {

    private val lifecycle = LifecycleRegistry(this)

    @field:Inject internal lateinit var view: LockListItemView
    @field:Inject internal lateinit var publisher: EventBus<LockListEvent>

    init {
      Injector.obtain<PadLockComponent>(itemView.context.applicationContext)
          .plusLockListItemComponent()
          .owner(this)
          .itemView(itemView)
          .build()
          .inject(this)
    }

    override fun getLifecycle(): Lifecycle = lifecycle

    fun bind(model: AppEntry) {
      view.bind(model)

      view.onSwitchChanged {
        Timber.d("Modify the database entry: ${model.packageName} $it")
        publisher.publish(LockListEvent(model.packageName, null, model.system, it))
      }

      lifecycle.fakeBind()
    }

    fun unbind() {
      // All the visible fields are explicitly set in bind so we don't need to unbind them
      // We do want to clear out the lifecycle for any async processes though.
      view.unbind()
      lifecycle.fakeUnbind()
    }

  }
}
